In [1]:
import pandas as pd

In [2]:
df = pd.read_csv("data/raw_wind.csv")
df.head()

C:\Users\audre\anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3146: DtypeWarning: Columns (21,27,28,29,30,31,33,38,42,44,50,52,56,72,89) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,


STATION                 DATE REPORT_TYPE  SOURCE  AWND  \
0  72392523190  2017-01-01T00:53:00       FM-15       7   NaN   
1  72392523190  2017-01-01T01:53:00       FM-15       7   NaN   
2  72392523190  2017-01-01T02:53:00       FM-15       7   NaN   
3  72392523190  2017-01-01T03:53:00       FM-15       7   NaN   
4  72392523190  2017-01-01T04:53:00       FM-15       7   NaN   

   BackupDirection  BackupDistance  BackupDistanceUnit  BackupElements  \
0              NaN             NaN                 NaN             NaN   
1              NaN             NaN                 NaN             NaN   
2              NaN             NaN                 NaN             NaN   
3              NaN             NaN                 NaN             NaN   
4              NaN             NaN                 NaN             NaN   

   BackupElevation  ...  ShortDurationPrecipitationValue045  \
0              NaN  ...                                 NaN   
1              NaN  ...                                 NaN   
2              NaN  ...                                 NaN   
3              NaN  ...                                 NaN   
4              NaN  ...                                 NaN   

   ShortDurationPrecipitationValue060  ShortDurationPrecipitationValue080  \
0                                 NaN                                 NaN   
1                                 NaN                                 NaN   
2                                 NaN                                 NaN   
3                                 NaN                                 NaN   
4                                 NaN                                 NaN   

   ShortDurationPrecipitationValue100  ShortDurationPrecipitationValue120  \
0                                 NaN                                 NaN   
1                                 NaN                                 NaN   
2                                 NaN                                 NaN   
3                                 NaN                                 NaN   
4                                 NaN                                 NaN   

   ShortDurationPrecipitationValue150  ShortDurationPrecipitationValue180  \
0                                 NaN                                 NaN   
1                                 NaN                                 NaN   
2                                 NaN                                 NaN   
3                                 NaN                                 NaN   
4                                 NaN                                 NaN   

   Sunrise  Sunset  WindEquipmentChangeDate  
0      NaN     NaN               2007-06-22  
1      NaN     NaN               2007-06-22  
2      NaN     NaN               2007-06-22  
3      NaN     NaN               2007-06-22  
4      NaN     NaN               2007-06-22  

[5 rows x 124 columns]

In [3]:
list(df.columns)

['STATION',
 'DATE',
 'REPORT_TYPE',
 'SOURCE',
 'AWND',
 'BackupDirection',
 'BackupDistance',
 'BackupDistanceUnit',
 'BackupElements',
 'BackupElevation',
 'BackupElevationUnit',
 'BackupEquipment',
 'BackupLatitude',
 'BackupLongitude',
 'BackupName',
 'CDSD',
 'CLDD',
 'DSNW',
 'DYHF',
 'DYTS',
 'DailyAverageDewPointTemperature',
 'DailyAverageDryBulbTemperature',
 'DailyAverageRelativeHumidity',
 'DailyAverageSeaLevelPressure',
 'DailyAverageStationPressure',
 'DailyAverageWetBulbTemperature',
 'DailyAverageWindSpeed',
 'DailyCoolingDegreeDays',
 'DailyDepartureFromNormalAverageTemperature',
 'DailyHeatingDegreeDays',
 'DailyMaximumDryBulbTemperature',
 'DailyMinimumDryBulbTemperature',
 'DailyPeakWindDirection',
 'DailyPeakWindSpeed',
 'DailyPrecipitation',
 'DailySnowDepth',
 'DailySnowfall',
 'DailySustainedWindDirection',
 'DailySustainedWindSpeed',
 'DailyWeather',
 'HDSD',
 'HTDD',
 'HourlyAltimeterSetting',
 'HourlyDewPointTemperature',
 'HourlyDryBulbTemperature',
 'Hourl

In [4]:
hour = df[['DATE','HourlyWindDirection','HourlyWindSpeed']]

In [5]:
hour.head()

DATE HourlyWindDirection  HourlyWindSpeed
0  2017-01-01T00:53:00                 050              3.0
1  2017-01-01T01:53:00                 000              0.0
2  2017-01-01T02:53:00                 000              0.0
3  2017-01-01T03:53:00                 020              7.0
4  2017-01-01T04:53:00                 000              0.0

In [6]:
hour.iloc[:,1].isna().to_list()

[False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 True,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 True,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 True,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 Fa

In [7]:
hour.iloc[:,2].isna().sum()

2022

In [8]:
hour[hour.iloc[:,1].isna().to_list()].iloc[:,2].isna().sum()

2022

This shows that all of the na values in the second column correspond with na values in the first column, so ok to delete since there's no info anyways. However, might be beneficial to keep for now just for merging purposes

**To Do:**
- find out what units the direction and speed are in
- find out what's going on with that date

In [9]:
hour['DATE'].head(26)

0     2017-01-01T00:53:00
1     2017-01-01T01:53:00
2     2017-01-01T02:53:00
3     2017-01-01T03:53:00
4     2017-01-01T04:53:00
5     2017-01-01T05:53:00
6     2017-01-01T06:53:00
7     2017-01-01T07:53:00
8     2017-01-01T08:53:00
9     2017-01-01T09:53:00
10    2017-01-01T10:53:00
11    2017-01-01T11:53:00
12    2017-01-01T12:53:00
13    2017-01-01T13:53:00
14    2017-01-01T14:53:00
15    2017-01-01T15:53:00
16    2017-01-01T16:53:00
17    2017-01-01T17:53:00
18    2017-01-01T18:53:00
19    2017-01-01T19:53:00
20    2017-01-01T20:53:00
21    2017-01-01T21:53:00
22    2017-01-01T22:53:00
23    2017-01-01T23:53:00
24    2017-01-01T23:59:00
25    2017-01-02T00:53:00
Name: DATE, dtype: object

According to the data's documentation (see google doc), wind speed is in mph and wind direction is in degrees to the left of north. Time on the other hand is a bit more complicated - it's in local standard time (no daylight savings) so must adjust to GMT/UTC+0 in order to standardize. However, on the 2021 end, this will result in a few missed hours of when our 2021 ended vs GMT's 2021....this actually isn't an issue becasue those hours would be dark anyways so not good surfing conditions heh

In [11]:
hour[hour.iloc[:,1].isna().to_list()]

DATE HourlyWindDirection  HourlyWindSpeed
24     2017-01-01T23:59:00                 NaN              NaN
54     2017-01-02T23:59:00                 NaN              NaN
82     2017-01-03T23:59:00                 NaN              NaN
144    2017-01-04T23:59:00                 NaN              NaN
187    2017-01-05T23:59:00                 NaN              NaN
...                    ...                 ...              ...
54591  2021-12-30T22:39:00                 NaN              NaN
54595  2021-12-30T23:59:00                 NaN              NaN
54619  2021-12-31T23:53:00                 NaN              NaN
54620  2021-12-31T23:59:00                 NaN              NaN
54621  2021-12-31T23:59:00                 NaN              NaN

[2031 rows x 3 columns]

Will clean up the na values when I have the entire dataset because I don't want to have merging issues

In [14]:
hour.shape

(54622, 3)

In [20]:
24*365*5+24

43824

**^This is the number of hours in the five years I'm looking at**

In [18]:
hour.drop_duplicates()

DATE HourlyWindDirection  HourlyWindSpeed
0      2017-01-01T00:53:00                 050              3.0
1      2017-01-01T01:53:00                 000              0.0
2      2017-01-01T02:53:00                 000              0.0
3      2017-01-01T03:53:00                 020              7.0
4      2017-01-01T04:53:00                 000              0.0
...                    ...                 ...              ...
54616  2021-12-31T20:53:00                 010              6.0
54617  2021-12-31T21:53:00                 330              6.0
54618  2021-12-31T22:53:00                 000              0.0
54619  2021-12-31T23:53:00                 NaN              NaN
54620  2021-12-31T23:59:00                 NaN              NaN

[54554 rows x 3 columns]

**Still has too many hours.....**

In [28]:
from datetime import timedelta

In [89]:
#converting PST to UTC
hour['DATE'] = pd.to_datetime(hour['DATE'], utc=True)+timedelta(hours=8)

<ipython-input-89-49c721bfeaca>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  hour['DATE'] = pd.to_datetime(hour['DATE'], utc=True)+timedelta(hours=8)


In [78]:
hour['DATE'] = hour['DATE']+timedelta(minutes=7)

<ipython-input-78-412ed0115dd4>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  hour['DATE'] = hour['DATE']+timedelta(minutes=7)


In [49]:
date = hour['DATE']+timedelta(minutes=7)

In [56]:
date.head()

0   2017-01-01 17:00:00
1   2017-01-01 18:00:00
2   2017-01-01 19:00:00
3   2017-01-01 20:00:00
4   2017-01-01 21:00:00
Name: DATE, dtype: datetime64[ns]

In [57]:
hour.head()

DATE HourlyWindDirection  HourlyWindSpeed
0 2017-01-01 16:53:00                 050              3.0
1 2017-01-01 17:53:00                 000              0.0
2 2017-01-01 18:53:00                 000              0.0
3 2017-01-01 19:53:00                 020              7.0
4 2017-01-01 20:53:00                 000              0.0

In [51]:
minutes = date.dt.minute != 0

In [52]:
minutes.sum()

10835

In [53]:
minutes.shape

(54622,)

In [58]:
hour[minutes].dropna()

DATE HourlyWindDirection  HourlyWindSpeed
39    2017-01-03 06:02:00                 150              7.0
40    2017-01-03 06:14:00                 140              6.0
41    2017-01-03 06:51:00                 180              5.0
43    2017-01-03 07:28:00                 230              6.0
51    2017-01-03 14:51:00                 000              0.0
...                   ...                 ...              ...
54581 2021-12-31 08:42:00                 010              5.0
54583 2021-12-31 09:51:00                 000              0.0
54586 2021-12-31 11:00:00                 000              0.0
54590 2021-12-31 14:18:00                 000              0.0
54592 2021-12-31 14:51:00                 070              5.0

[8933 rows x 3 columns]

In [62]:
hour[[not x for x in minutes]]

DATE HourlyWindDirection  HourlyWindSpeed
0     2017-01-01 16:53:00                 050              3.0
1     2017-01-01 17:53:00                 000              0.0
2     2017-01-01 18:53:00                 000              0.0
3     2017-01-01 19:53:00                 020              7.0
4     2017-01-01 20:53:00                 000              0.0
...                   ...                 ...              ...
54615 2022-01-01 11:53:00                 280              6.0
54616 2022-01-01 12:53:00                 010              6.0
54617 2022-01-01 13:53:00                 330              6.0
54618 2022-01-01 14:53:00                 000              0.0
54619 2022-01-01 15:53:00                 NaN              NaN

[43787 rows x 3 columns]

In [63]:
43787-(24*365*5+24)

-37

Missing 37 hours of data if I just drop all of the entries that don't fall right on the hour.....that's not too bad given that 

In [66]:
(37*2/43824)*100

0.16885724717050019

it's only 0.17% of availiable data....

Since there are likely data points in the entire wind data set for every hour I care about, if this were higher stakes, I could figure out ways to extract that information and substitute it in for the 37 hours of missing data. However, since 37 hours is such a small percentage of approximate daylight hours, I'm just going to toss them.

In [73]:
hour['DATE'][0].tzinfo.utcoffset(None)

datetime.timedelta(0)

In [83]:
hour[hour['DATE'].dt.minute != 0]

DATE HourlyWindDirection  HourlyWindSpeed
24    2017-01-03 00:06:00+00:00                 NaN              NaN
39    2017-01-03 14:09:00+00:00                 150              7.0
40    2017-01-03 14:21:00+00:00                 140              6.0
41    2017-01-03 14:58:00+00:00                 180              5.0
43    2017-01-03 15:35:00+00:00                 230              6.0
...                         ...                 ...              ...
54591 2021-12-31 22:46:00+00:00                 NaN              NaN
54592 2021-12-31 22:58:00+00:00                 070              5.0
54595 2022-01-01 00:06:00+00:00                 NaN              NaN
54620 2022-01-02 00:06:00+00:00                 NaN              NaN
54621 2022-01-02 00:06:00+00:00                 NaN              NaN

[10835 rows x 3 columns]

In [98]:
hour[hour['DATE'].dt.minute == 0]['DATE'].drop_duplicates().dt.year.value_counts()

2020    8780
2021    8759
2019    8750
2018    8743
2017    8721
2022      33
Name: DATE, dtype: int64

In [96]:
hour['DATE'].dt.year.value_counts()

2020    11014
2017    11003
2021    10933
2019    10897
2018    10733
2022       42
Name: DATE, dtype: int64

In [97]:
hour[hour['DATE'].dt.minute != 0]['DATE'].dt.year.value_counts()

2017    2282
2020    2234
2021    2174
2019    2146
2018    1990
2022       9
Name: DATE, dtype: int64

In [100]:
hour[hour['DATE'].dt.minute == 0].isna().sum()

DATE                     0
HourlyWindDirection    129
HourlyWindSpeed        124
dtype: int64

In [101]:
hour[hour['DATE'].dt.minute != 0].isna().sum()

DATE                      0
HourlyWindDirection    1902
HourlyWindSpeed        1898
dtype: int64

In [103]:
hour[hour['DATE'].dt.minute != 0].shape

(10835, 3)

Sooooooo there is a lot of missing data in the weird-timed measurements, almost 20% is NaN values.... so once again probably not worth it to try and suppliment top of the hour data even though the 129 NaN values make it a bit worse than just 37 missing hours....but, as long as those missing values are nestled betweeen hours with data, I can take an average or smth of the two hours with data to fill those blanks. THEN would maybe not even have 37 missing hours after all

In [164]:
top = hour[hour['DATE'].dt.minute ==0].reset_index()

In [184]:
#getting list of na indicies so I can create an average to fill them
nalist = top[top['HourlyWindDirection'].isna()].index.tolist()

In [166]:
#getting list of na indicies so I can create an average to fill them
nalist = top[top['HourlyWindSpeed'].isna()].index.tolist()

In [170]:
nalist[10:20]

[2634, 2798, 3178, 4152, 4613, 5531, 6035, 6241, 6242, 6243]

In [185]:
#want to check if each na value is nestled between filled values
#can manually look at first and last values
badna=[]
for i in range(1,len(nalist)-1):
    if (nalist[i]+1 == nalist[i+1]):
        badna.append(nalist[i])
    if (nalist[i]-1==nalist[i-1]):
        badna.append(nalist[i])

In [205]:
len(badna)

16

I think I'll just have to chill with these 12 nan values, not that big of a deal, don't have to throw entire hour out of full dataset, can just not say that the wind was perfect during these hours

In [149]:
nalist[-5:]

[43283, 43557, 43620, 43719, 43786]

Now I have to figure out how to fill those NaN values with an average of the two entries it's between.
I'll do wind speed first

In [153]:
top.tail()

index                      DATE HourlyWindDirection  HourlyWindSpeed
43782  54615 2022-01-02 04:00:00+00:00                 280              6.0
43783  54616 2022-01-02 05:00:00+00:00                 010              6.0
43784  54617 2022-01-02 06:00:00+00:00                 330              6.0
43785  54618 2022-01-02 07:00:00+00:00                 000              0.0
43786  54619 2022-01-02 08:00:00+00:00                 NaN              NaN

Last entry is NaN so can't do the average thing....but will be chucking this hour anyways cuz it's midnight on new years, no one surfs then

In [154]:
#getting list of na indicies so I can create an average to fill them
naspeed = top[top['HourlyWindSpeed'].isna()].index.tolist()

naspeed.remove(43786)

#making a copy of wind speed column that I'll edit
speed = top['HourlyWindSpeed']
for i in naspeed:
    speed[i]=(speed[i-1]+speed[i+1])/2

In [176]:
top['Wind Speed']=speed
top.head()

index                      DATE HourlyWindDirection  HourlyWindSpeed  \
0      0 2017-01-02 09:00:00+00:00                 050              3.0   
1      1 2017-01-02 10:00:00+00:00                 000              0.0   
2      2 2017-01-02 11:00:00+00:00                 000              0.0   
3      3 2017-01-02 12:00:00+00:00                 020              7.0   
4      4 2017-01-02 13:00:00+00:00                 000              0.0   

   Wind Speed  
0         3.0  
1         0.0  
2         0.0  
3         7.0  
4         0.0

In [177]:
top['Wind Speed'].isna().sum()

11

In [178]:
len(naspeed)

123

11 missing hours is a lot better than 123 missing hours. Now for direction

In [179]:
top['HourlyWindDirection'].dtype

dtype('O')

Have to convert wind direction to integers in order to do average

In [180]:
top['HourlyWindDirection'].value_counts()

000    11961
260     2458
250     2213
150     1719
VRB     1669
140     1654
240     1541
270     1436
130     1205
160     1195
120     1018
230      939
090      895
080      882
110      830
220      743
100      740
070      722
280      710
050      688
170      681
040      667
060      651
210      629
200      591
020      517
290      511
030      495
010      454
190      412
300      403
310      395
180      384
320      358
360      344
330      332
340      320
350      296
Name: HourlyWindDirection, dtype: int64

```VRB``` and ```000``` do not measure a wind direction....```VRB``` means too variable to say, and ```000``` means calm. So I'll just take them out for the average calculation.

In [187]:
direction = top['HourlyWindDirection']
d_index = top['HourlyWindDirection'].index.to_list()
calm_winds = [i for i in d_index if direction[i] in ('VRB', '000')]

In [189]:
direction[calm_winds]

1        000
2        000
4        000
8        000
9        000
        ... 
43765    000
43768    000
43770    000
43781    VRB
43785    000
Name: HourlyWindDirection, Length: 13630, dtype: object

In [192]:
nadirection = direction[direction.isna()].index.to_list()

int_direction = direction.drop(index=calm_winds+nadirection).astype(int)

In [196]:
nadirection.remove(43786)

In [197]:
int_direction.head()

0     50
3     20
5     80
6    260
7     40
Name: HourlyWindDirection, dtype: int32

In [198]:
direction.iloc[int_direction.index]=int_direction

C:\Users\audre\anaconda3\lib\site-packages\pandas\core\indexing.py:670: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  iloc._setitem_with_indexer(indexer, value)


In [202]:
type(direction[0])

int

In [211]:
for i in nadirection:
    if (type(direction[i-1])==int) & (type(direction[i+1])==int):
         direction[i]=(direction[i-1]+direction[i+1])/2

<ipython-input-211-99ad86744759>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  direction[i]=(direction[i-1]+direction[i+1])/2


In [206]:
len(nadirection)

128

In [207]:
len(badna)

16

Hopefully this is less than 128. Must be 16 or more though

In [212]:
direction.isna().sum()

45

ok we can work with that, much better than before

In [213]:
top['Wind Direction']=direction

In [214]:
top.head()

index                      DATE HourlyWindDirection  HourlyWindSpeed  \
0      0 2017-01-02 09:00:00+00:00                  50              3.0   
1      1 2017-01-02 10:00:00+00:00                 000              0.0   
2      2 2017-01-02 11:00:00+00:00                 000              0.0   
3      3 2017-01-02 12:00:00+00:00                  20              7.0   
4      4 2017-01-02 13:00:00+00:00                 000              0.0   

   Wind Speed Wind Direction  
0         3.0             50  
1         0.0            000  
2         0.0            000  
3         7.0             20  
4         0.0            000

In [223]:
top.iloc[:,[4,5]].isna().sum()

Wind Speed        11
Wind Direction    45
dtype: int64

Much better than before :)